provider/boltmem/boltmem.go

```diff
@@ -0,0 +1,426 @@
+package boltmem
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/boltdb/bolt"
+	"github.com/prometheus/alertmanager/provider"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+)
+
+var (
+	bktNotifies = []byte("notifies")
+	bktSilences = []byte("silences")
+	bktAlerts   = []byte("alerts")
+)
+
+// Alerts gives access to a set of alerts. All methods are goroutine-safe.
+type Alerts struct {
+	db *bolt.DB
+
+	mtx       sync.RWMutex
+	listeners map[int]chan *types.Alert
+	next      int
+}
+
+func NewAlerts(path string) (*Alerts, error) {
+	db, err := bolt.Open(filepath.Join(path, "alerts.db"), 0666, nil)
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(bktAlerts)
+		return err
+	})
+	return &Alerts{
+		db:        db,
+		listeners: map[int]chan *types.Alert{},
+		next:      0,
+	}, err
+}
+
+func (a *Alerts) Close() error {
+	return a.db.Close()
+}
+
+// Subscribe returns an iterator over active alerts that have not been
+// resolved and successfully notified about.
+// They are not guaranteed to be in chronological order.
+func (a *Alerts) Subscribe() provider.AlertIterator {
+	var (
+		ch   = make(chan *types.Alert, 200)
+		done = make(chan struct{})
+	)
+	alerts, err := a.getPending()
+
+	i := a.next
+	a.next++
+
+	a.mtx.Lock()
+	a.listeners[i] = ch
+	a.mtx.Unlock()
+
+	go func() {
+		defer func() {
+			a.mtx.Lock()
+			delete(a.listeners, i)
+			close(ch)
+			a.mtx.Unlock()
+		}()
+
+		for _, a := range alerts {
+			select {
+			case ch <- a:
+			case <-done:
+				return
+			}
+		}
+
+		<-done
```

Comment:
Here you have an extra `<-done` to wait with the cleanup until `done` is closed, not just if `a` is closed. But in the method below you don't. Is there a reason that makes it necessary to have this here?